{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f82e56ee",
   "metadata": {},
   "source": [
    "## 1. Download the datasets from UCSC website"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a6562c1",
   "metadata": {},
   "source": [
    "### 1.1 Download the multi-omics data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32751c87",
   "metadata": {},
   "source": [
    "* Parse the data from the UCSC Xena website in PANCAN cohort:\n",
    "https://xenabrowser.net/datapages/?cohort=TCGA%20Pan-Cancer%20(PANCAN)&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443\n",
    "\n",
    "* Copy number (gene-level) - gene-level copy number (gistic2_thresholded)\n",
    "    * Dataset: https://xenabrowser.net/datapages/?dataset=TCGA.PANCAN.sampleMap%2FGistic2_CopyNumber_Gistic2_all_thresholded.by_genes&host=https%3A%2F%2Ftcga.xenahubs.net&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443\n",
    "\n",
    "* DNA methylation (Methylation450K)\n",
    "    * Dataset: https://xenabrowser.net/datapages/?dataset=jhu-usc.edu_PANCAN_HumanMethylation450.betaValue_whitelisted.tsv.synapse_download_5096262.xena&host=https%3A%2F%2Fpancanatlas.xenahubs.net&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443\n",
    "    * ID Map: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GPL16304\n",
    "\n",
    "* Gene expression RNAseq - TOIL RSEM fpkm\n",
    "    * Dataset: https://xenabrowser.net/datapages/?dataset=tcga_RSEM_gene_fpkm&host=https%3A%2F%2Ftoil.xenahubs.net&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443\n",
    "    \n",
    "* Protein expression - RPPA\n",
    "    * Dataset: https://xenabrowser.net/datapages/?dataset=TCGA-RPPA-pancan-clean.xena&host=https%3A%2F%2Fpancanatlas.xenahubs.net&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "05f98026",
   "metadata": {},
   "source": [
    "### 1.2 Download the clinical data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1196e4a",
   "metadata": {},
   "source": [
    "* Phenotype - Curated clinical data\n",
    "    * Dataset: https://xenabrowser.net/datapages/?dataset=Survival_SupplementalTable_S1_20171025_xena_sp&host=https%3A%2F%2Fpancanatlas.xenahubs.net&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443\n",
    "\n",
    "* Phenotype - Immune subtype\n",
    "    * Dataset: https://xenabrowser.net/datapages/?dataset=Subtype_Immune_Model_Based.txt&host=https%3A%2F%2Fpancanatlas.xenahubs.net&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443\n",
    "\n",
    "* Phenotype - Molecular subtype\n",
    "    * Dataset: https://xenabrowser.net/datapages/?dataset=TCGASubtype.20170308.tsv&host=https%3A%2F%2Fpancanatlas.xenahubs.net&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443\n",
    "\n",
    "* Phenotype - sample type and primary disease\n",
    "    * Dataset: https://xenabrowser.net/datapages/?dataset=TCGA_phenotype_denseDataOnlyDownload.tsv&host=https%3A%2F%2Fpancanatlas.xenahubs.net&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c013cb4",
   "metadata": {},
   "source": [
    "## 2. Read the files"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e17ed47",
   "metadata": {},
   "source": [
    "### 2.1 Read DNA methylation data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7b13f2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "methylation_value = pd.read_csv('./UCSC-raw/jhu-usc.edu_PANCAN_HumanMethylation450.betaValue_whitelisted.tsv.synapse_download_5096262.xena', delimiter='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d15bd3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "methylation_value"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c2fd3d8",
   "metadata": {},
   "source": [
    "### 2.2 Read Platform annotations for 450k methylation "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5144755",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "# Reading and processing basic data\n",
    "try:\n",
    "    annotation = pd.read_table('./UCSC-raw/GPL16304-47833.txt', delimiter='\\t')\n",
    "    annotation['Distance_closest_TSS'] = annotation['Distance_closest_TSS'].astype(int)\n",
    "    annotation = annotation[~annotation['Closest_TSS'].apply(lambda x: len(str(x).split(';')) > 1)]\n",
    "except ValueError as e:\n",
    "    print(f\"Unable to convert 'Closest_TSS' column to integer: {e}\")\n",
    "    problematic_rows = annotation['Distance_closest_TSS'].apply(lambda x: not str(x).isnumeric())\n",
    "    print(\"Problematic rows:\")\n",
    "    print(annotation.loc[problematic_rows])\n",
    "\n",
    "annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "795ee450",
   "metadata": {},
   "outputs": [],
   "source": [
    "map_annotation= annotation[['ID', 'Closest_TSS','Closest_TSS_gene_name', 'Distance_closest_TSS']]\n",
    "map_annotation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06fd142d",
   "metadata": {},
   "source": [
    "### 2.3 Read copynumber data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "659fb78a",
   "metadata": {},
   "outputs": [],
   "source": [
    "copynumber = pd.read_csv('./UCSC-raw/Gistic2_CopyNumber_Gistic2_all_thresholded.by_genes',sep='\\t')\n",
    "copynumber"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b5ca523",
   "metadata": {},
   "source": [
    "### 2.4 Read gene expression data, mapping information and substitue the the gene name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "988b58cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_expression = pd.read_csv('./UCSC-raw/tcga_RSEM_gene_fpkm', sep='\\t')\n",
    "gene_expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5021d202",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_expression_map = pd.read_csv('./UCSC-raw/probeMap_gencode.v23.annotation.gene.probemap', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79a01371",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_expression_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "884042e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "expression_merged = pd.merge(gene_expression, gene_expression_map, left_on='sample', right_on='id', how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7b5d069",
   "metadata": {},
   "outputs": [],
   "source": [
    "expression_merged.drop(columns=['sample','chrom', 'chromStart','chromEnd','strand','id'], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7667c3f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# set gene to the first column\n",
    "cols = ['gene'] + [col for col in expression_merged if col != 'gene']\n",
    "gene_expression = expression_merged[cols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ba98336",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_expression "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7dcfd8c8",
   "metadata": {},
   "source": [
    "### 2.5 Read clinical data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fda02764",
   "metadata": {},
   "outputs": [],
   "source": [
    "survival = pd.read_csv('./UCSC-raw/Survival_SupplementalTable_S1_20171025_xena_sp', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f3fb665",
   "metadata": {},
   "outputs": [],
   "source": [
    "survival"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59ef08bd",
   "metadata": {},
   "source": [
    "### 2.6 Read immune subtype data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0eb81830",
   "metadata": {},
   "outputs": [],
   "source": [
    "immune_subtype = pd.read_csv('./UCSC-raw/Subtype_Immune_Model_Based.txt',sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28b83c94",
   "metadata": {},
   "outputs": [],
   "source": [
    "immune_subtype"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a572ec1e",
   "metadata": {},
   "source": [
    "### 2.7 Read proteomics data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "616e7abc",
   "metadata": {},
   "outputs": [],
   "source": [
    "protein = pd.read_csv('./UCSC-raw/TCGA-RPPA-pancan-clean.xena',sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d4385da",
   "metadata": {},
   "outputs": [],
   "source": [
    "protein"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43152c82",
   "metadata": {},
   "source": [
    "### 2.8 Read molecular subtype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4aebe0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "cellsub = pd.read_csv('./UCSC-raw/TCGASubtype.20170308.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "434cf0fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "cellsub"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15538083",
   "metadata": {},
   "source": [
    "### 2.9 Read sample type and primary disease"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec3af07c",
   "metadata": {},
   "outputs": [],
   "source": [
    "dense = pd.read_csv('./UCSC-raw/TCGA_phenotype_denseDataOnlyDownload.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e0c927c",
   "metadata": {},
   "outputs": [],
   "source": [
    "dense"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f0c2026",
   "metadata": {},
   "source": [
    "## 3. Methylation data process"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e23e6e4",
   "metadata": {},
   "source": [
    "### 3.1 Define methylation region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d62d283a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "#Define vectorized area determination function for methylation data\n",
    "def vectorized_determine_region(distances):\n",
    "    regions = ['Upstream', 'Distal Promoter', 'Proximal Promoter', 'Core Promoter', 'Downstream']\n",
    "    conditions = [\n",
    "        (-6000 <= distances) & (distances < -3000),\n",
    "        (-3000 <= distances) & (distances < -250),\n",
    "        (-250 <= distances) & (distances < -50),\n",
    "        (-50 <= distances) & (distances <= 0),\n",
    "        (0 < distances) & (distances <= 3000)\n",
    "    ]\n",
    "    return np.select(conditions, regions, default=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38b53554",
   "metadata": {},
   "source": [
    "### 3.2 Merge the annotation files to the methylation data and apply region function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e90a582",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merging basic data and methylation data\n",
    "methylation_merged_df = pd.merge(map_annotation, methylation_value, left_on='ID', right_on='sample', how='right')\n",
    "\n",
    "# Determining the region for each row outside the loop\n",
    "methylation_merged_df['Region'] = vectorized_determine_region(methylation_merged_df['Distance_closest_TSS'])\n",
    "\n",
    "methylation_merged_df = methylation_merged_df.dropna(subset=['Region'])  # Remove rows without a region\n",
    "\n",
    "# Initializing a dictionary to store data for each region\n",
    "regions_data = {region: pd.DataFrame() for region in [\"Upstream\", \"Distal Promoter\", \"Proximal Promoter\", \"Core Promoter\", \"Downstream\"]}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d03483e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "methylation_merged_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7ebd64c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Delete the 'sample' column\n",
    "methylation_merged_df = methylation_merged_df.drop('sample', axis=1)\n",
    "# Delete the 'ID' column\n",
    "methylation_merged_df = methylation_merged_df.drop('ID', axis=1)\n",
    "# Delete the 'Distance_closest_TSS' column\n",
    "methylation_merged_df = methylation_merged_df.drop('Distance_closest_TSS', axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b7650c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "methylation_merged_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2459ff7",
   "metadata": {},
   "outputs": [],
   "source": [
    "methylation_merged_df['Closest_TSS'] = methylation_merged_df['Closest_TSS'].astype(int)\n",
    "methylation_merged_df['Closest_TSS_gene_name'] = methylation_merged_df['Closest_TSS_gene_name'].astype(str)\n",
    "methylation_merged_df['Region'] = methylation_merged_df['Region'].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac9803f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(methylation_merged_df[['Closest_TSS', 'Closest_TSS_gene_name', 'Region']].dtypes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6df49e8",
   "metadata": {},
   "source": [
    "### 3.3 Calculate the average methylation value of five regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17e307b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# find columns not started with 'TCGA'\n",
    "non_tcga_columns = methylation_merged_df.filter(regex='^(?!TCGA)').columns\n",
    "\n",
    "print(\"Columns not starting with TCGA:\")\n",
    "print(non_tcga_columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abc89f14",
   "metadata": {},
   "outputs": [],
   "source": [
    "# obtain all regions\n",
    "regions = methylation_merged_df['Region'].unique()\n",
    "regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7a92755",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "# Initialize empty DataFrames for each region\n",
    "Upstream_df = pd.DataFrame()\n",
    "Distal_Promoter_df = pd.DataFrame()\n",
    "Proximal_Promoter_df = pd.DataFrame()\n",
    "Core_Promoter_df = pd.DataFrame()\n",
    "Downstream_df = pd.DataFrame()\n",
    "\n",
    "# Operate on each region\n",
    "for region in regions:\n",
    "    # Get all data for this region\n",
    "    region_data = methylation_merged_df[methylation_merged_df['Region'] == region]\n",
    "    \n",
    "    # Group and calculate the average for each (TSS, Region) combination\n",
    "    grouped = region_data.groupby(['Closest_TSS_gene_name', 'Region'], as_index=False).mean()\n",
    "    \n",
    "    # Since we split the data into different files based on Region, we can delete this column\n",
    "    grouped = grouped.drop(columns=['Region'])\n",
    "    \n",
    "    # Print the shape of the grouped data\n",
    "    print(f\"Shape of {region}: {grouped.shape}\")\n",
    "    \n",
    "    # Assign the grouped data to the respective DataFrame\n",
    "    if region == 'Upstream':\n",
    "        Upstream_df = grouped\n",
    "    elif region == 'Distal Promoter':\n",
    "        Distal_Promoter_df = grouped\n",
    "    elif region == 'Proximal Promoter':\n",
    "        Proximal_Promoter_df = grouped\n",
    "    elif region == 'Core Promoter':\n",
    "        Core_Promoter_df = grouped\n",
    "    elif region == 'Downstream':\n",
    "        Downstream_df = grouped\n",
    "\n",
    "    # Optionally, save the data for this region to a new csv file\n",
    "    # grouped.to_csv(f\"{region}_averaged_tss_data.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b44016c9",
   "metadata": {},
   "source": [
    "### 3.4 Unify gene and TSS for five methylation value files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a23c454",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# from methylation files above DataFrame \n",
    "dfs = [Upstream_df, Distal_Promoter_df, Proximal_Promoter_df, Core_Promoter_df, Downstream_df]\n",
    "\n",
    "# merge those files to find all combos \n",
    "all_genes_tss = pd.concat(dfs)['Closest_TSS_gene_name'].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8116e5da",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_genes_tss"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c15e8a3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge unique combinations back into each DataFrame and fill NaN values with 0\n",
    "# Upstream\n",
    "Upstream_df = pd.merge(all_genes_tss, Upstream_df, on=['Closest_TSS_gene_name'], how='outer').fillna(0)\n",
    "print(f\"Shape of Upstream_df: {Upstream_df.shape}\")\n",
    "\n",
    "# Distal Promoter\n",
    "Distal_Promoter_df = pd.merge(all_genes_tss, Distal_Promoter_df, on=['Closest_TSS_gene_name'], how='outer').fillna(0)\n",
    "print(f\"Shape of Distal_Promoter_df: {Distal_Promoter_df.shape}\")\n",
    "\n",
    "# Proximal Promoter\n",
    "Proximal_Promoter_df = pd.merge(all_genes_tss, Proximal_Promoter_df, on=['Closest_TSS_gene_name'], how='outer').fillna(0)\n",
    "print(f\"Shape of Proximal_Promoter_df: {Proximal_Promoter_df.shape}\")\n",
    "\n",
    "# Core Promoter\n",
    "Core_Promoter_df = pd.merge(all_genes_tss, Core_Promoter_df, on=['Closest_TSS_gene_name'], how='outer').fillna(0)\n",
    "print(f\"Shape of Core_Promoter_df: {Core_Promoter_df.shape}\")\n",
    "\n",
    "# Downstream\n",
    "Downstream_df = pd.merge(all_genes_tss, Downstream_df, on=['Closest_TSS_gene_name'], how='outer').fillna(0)\n",
    "print(f\"Shape of Downstream_df: {Downstream_df.shape}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1cec3601",
   "metadata": {},
   "source": [
    "## 4. Unify genes and patient samples within datasets"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9ff09f0",
   "metadata": {},
   "source": [
    "### 4.1 Unify gene and TSS for methylation, copynumer, and gene expression data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65df6203",
   "metadata": {},
   "outputs": [],
   "source": [
    "copynumber"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "028d812d",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9989681c",
   "metadata": {},
   "outputs": [],
   "source": [
    "Upstream_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86fb3187",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Convert the gene name columns from each DataFrame to sets\n",
    "copynumber_genes = set(copynumber['Sample'])\n",
    "gene_expression_genes = set(gene_expression['gene'])\n",
    "methylation_genes = set(Upstream_df['Closest_TSS_gene_name'])\n",
    "\n",
    "# Find the intersection of the three sets\n",
    "common_genes = copynumber_genes & gene_expression_genes & methylation_genes\n",
    "\n",
    "# Convert the intersection back to a list, if needed\n",
    "common_genes_list = list(common_genes)\n",
    "\n",
    "# Print the number of common genes\n",
    "print(f\"Number of common genes: {len(common_genes)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### 4.1.1 Intersecting genes with various databases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "# Add the gene names from databases like [KEGG / BioGRID] to intersect with the common genes\n",
    "# KEGG\n",
    "kegg_pathway_df = pd.read_csv('./Regulatory-network-data/KEGG/full_kegg_pathway_list.csv')\n",
    "kegg_pathway_df = kegg_pathway_df[['source', 'target', 'pathway_name']]\n",
    "kegg_df = kegg_pathway_df[kegg_pathway_df['pathway_name'].str.contains('signaling pathway|signaling pathways', case=False)]\n",
    "print(kegg_df['pathway_name'].value_counts())\n",
    "kegg_df = kegg_df.rename(columns={'source': 'src', 'target': 'dest'})\n",
    "src_list = list(kegg_df['src'])\n",
    "dest_list = list(kegg_df['dest'])\n",
    "path_list = list(kegg_df['pathway_name'])\n",
    "# ADJUST ALL GENES TO UPPERCASE\n",
    "up_src_list = []\n",
    "for src in src_list:\n",
    "    up_src = src.upper()\n",
    "    up_src_list.append(up_src)\n",
    "up_dest_list = []\n",
    "for dest in dest_list:\n",
    "    up_dest = dest.upper()\n",
    "    up_dest_list.append(up_dest)\n",
    "up_kegg_conn_dict = {'src': up_src_list, 'dest': up_dest_list}\n",
    "up_kegg_df = pd.DataFrame(up_kegg_conn_dict)\n",
    "up_kegg_df = up_kegg_df.drop_duplicates()\n",
    "up_kegg_df.to_csv('./Regulatory-network-data/KEGG/up_kegg.csv', index=False, header=True)\n",
    "kegg_gene_list = list(set(list(up_kegg_df['src']) + list(up_kegg_df['dest'])))\n",
    "print('----- NUMBER OF GENES IN KEGG: ' + str(len(kegg_gene_list)) + ' -----')\n",
    "print(up_kegg_df.shape)\n",
    "\n",
    "up_kegg_path_conn_dict = {'src': up_src_list, 'dest': up_dest_list, 'path': path_list}\n",
    "up_kegg_path_df = pd.DataFrame(up_kegg_path_conn_dict)\n",
    "up_kegg_path_df = up_kegg_path_df.drop_duplicates()\n",
    "up_kegg_path_df.to_csv('./Regulatory-network-data/KEGG/up_kegg_path.csv', index=False, header=True)\n",
    "kegg_path_gene_list = list(set(list(up_kegg_path_df['src']) + list(up_kegg_path_df['dest'])))\n",
    "print('----- NUMBER OF GENES IN KEGG PATH: ' + str(len(kegg_path_gene_list)) + ' -----')\n",
    "print(up_kegg_path_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# BioGRID\n",
    "biogrid_df = pd.read_table('./Regulatory-network-data/BioGrid/BIOGRID-ALL-3.5.174.mitab.Symbol.txt', delimiter = '\\t')\n",
    "eh_list = list(biogrid_df['e_h'])\n",
    "et_list = list(biogrid_df['e_t'])\n",
    "# ADJUST ALL GENES TO UPPERCASE\n",
    "up_eh_list = []\n",
    "for eh in eh_list:\n",
    "    up_eh = eh.upper()\n",
    "    up_eh_list.append(up_eh)\n",
    "up_et_list = []\n",
    "for et in et_list:\n",
    "    up_et = et.upper()\n",
    "    up_et_list.append(up_et)\n",
    "up_biogrid_conn_dict = {'e_h': up_eh_list, 'e_t': up_et_list}\n",
    "up_biogrid_df = pd.DataFrame(up_biogrid_conn_dict)\n",
    "print(up_biogrid_df)\n",
    "print(up_biogrid_df.shape)\n",
    "up_biogrid_df.to_csv('./Regulatory-network-data/BioGrid/up_biogrid.csv', index = False, header = True)\n",
    "up_biogrid_gene_list = list(set(list(up_biogrid_df['e_h']) + list(up_biogrid_df['e_t'])))\n",
    "print('----- NUMBER OF GENES IN BioGRID: ' + str(len(up_biogrid_gene_list)) + ' -----')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# STRING\n",
    "string_df = pd.read_csv('./Regulatory-network-data/STRING/9606.protein.links.detailed.v11.0_sym.csv', low_memory=False)\n",
    "src_list = list(string_df['Source'])\n",
    "tar_list = list(string_df['Target'])\n",
    "# ADJUST ALL GENES TO UPPERCASE\n",
    "up_src_list = []\n",
    "for src in src_list:\n",
    "    up_src = src.upper()\n",
    "    up_src_list.append(up_src)\n",
    "up_tar_list = []\n",
    "for tar in tar_list:\n",
    "    up_tar = tar.upper()\n",
    "    up_tar_list.append(up_tar)\n",
    "up_string_conn_dict = {'Source': up_src_list, 'Target': up_tar_list}\n",
    "up_string_df = pd.DataFrame(up_string_conn_dict)\n",
    "print(up_string_df)\n",
    "up_string_df.to_csv('./Regulatory-network-data/STRING/up_string.csv', index = False, header = True)\n",
    "up_string_gene_list = list(set(list(up_string_df['Source']) + list(up_string_df['Target'])))\n",
    "print('----- NUMBER OF GENES IN STRING: ' + str(len(up_string_gene_list)) + ' -----')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# intersect the [common genes] with the genes in the different databases [KEGG / BioGRID / STRING]\n",
    "selected_database = 'KEGG'\n",
    "# selected_database = 'BioGRID'\n",
    "# selected_database = 'STRING'\n",
    "if selected_database == 'KEGG':\n",
    "    common_genes = list(set(common_genes) & set(kegg_gene_list))\n",
    "    print('----- NUMBER OF INTERSECTED GENES IN KEGG: ' + str(len(common_genes)) + ' -----')\n",
    "elif selected_database == 'BioGRID':\n",
    "    common_genes = list(set(common_genes) & set(up_biogrid_gene_list))\n",
    "    print('----- NUMBER OF INTERSECTED GENES IN BioGRID: ' + str(len(common_genes)) + ' -----')\n",
    "elif selected_database == 'STRING':\n",
    "    common_genes = list(set(common_genes) & set(up_string_gene_list))\n",
    "    print('----- NUMBER OF INTERSECTED GENES IN STRING: ' + str(len(common_genes)) + ' -----')\n",
    "\n",
    "# filter the genes in the different databases [KEGG / BioGRID / STRING] with the [common genes]\n",
    "if selected_database == 'KEGG':\n",
    "    filtered_up_kegg_df = up_kegg_df[up_kegg_df['src'].isin(common_genes) & up_kegg_df['dest'].isin(common_genes)]\n",
    "    filtered_up_kegg_df = filtered_up_kegg_df.drop_duplicates()\n",
    "    filtered_up_kegg_df = filtered_up_kegg_df.sort_values(by=['src', 'dest']).reset_index(drop=True)\n",
    "    print('----- NEW KEGG EDGE CONNECTIONS: ' + str(len(filtered_up_kegg_df)) + ' -----')\n",
    "    filtered_up_kegg_path_df = up_kegg_path_df[up_kegg_path_df['src'].isin(common_genes) & up_kegg_path_df['dest'].isin(common_genes)]\n",
    "    filtered_up_kegg_path_df = filtered_up_kegg_path_df.drop_duplicates()\n",
    "    filtered_up_kegg_path_df = filtered_up_kegg_path_df.sort_values(by=['src', 'dest']).reset_index(drop=True)\n",
    "    print('----- NEW KEGG PATHWAY CONNECTIONS: ' + str(len(filtered_up_kegg_path_df)) + ' -----')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "display(filtered_up_kegg_df)\n",
    "display(filtered_up_kegg_path_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76578188",
   "metadata": {},
   "outputs": [],
   "source": [
    "# select common genes in copynumber data\n",
    "copynumber_filtered = copynumber.loc[copynumber['Sample'].isin(common_genes)]\n",
    "\n",
    "# select common genes in gene expression data\n",
    "gene_expression_filtered = gene_expression.loc[gene_expression['gene'].isin(common_genes)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99d70508",
   "metadata": {},
   "outputs": [],
   "source": [
    "copynumber_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4af488da",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_expression_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68ba62d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_expression_filtered['gene'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52038d35",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_expression_filtered = gene_expression_filtered.groupby('gene', as_index=False).mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5fab2f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_expression_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efcd71ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "# select common genes in methylation data\n",
    "Upstream_df_filtered = Upstream_df.loc[Upstream_df['Closest_TSS_gene_name'].isin(common_genes)]\n",
    "Distal_Promoter_df_filtered = Distal_Promoter_df.loc[Distal_Promoter_df['Closest_TSS_gene_name'].isin(common_genes)]\n",
    "Proximal_Promoter_df_filtered = Proximal_Promoter_df.loc[Proximal_Promoter_df['Closest_TSS_gene_name'].isin(common_genes)]\n",
    "Core_Promoter_df_filtered = Core_Promoter_df.loc[Core_Promoter_df['Closest_TSS_gene_name'].isin(common_genes)]\n",
    "Downstream_df_filtered = Downstream_df.loc[Downstream_df['Closest_TSS_gene_name'].isin(common_genes)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5459361",
   "metadata": {},
   "outputs": [],
   "source": [
    "Upstream_df_filtered"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ced5f75",
   "metadata": {},
   "source": [
    "### 4.2 Unify patient samples within methylation, copynumer,  gene expression, clinical, proteomics, molecular subtype, and sample type, primary disease datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2063f64a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# clinical data\n",
    "survival"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e456736",
   "metadata": {},
   "outputs": [],
   "source": [
    "#immune subtype\n",
    "immune_subtype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4bac904",
   "metadata": {},
   "outputs": [],
   "source": [
    "#proteomics data\n",
    "protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b5ce47c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#molecular subtype\n",
    "cellsub"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f2f995d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#sample type and primary disease\n",
    "dense"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3922af8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Upstream_df_filtered, Distal_Promoter_df_filtered, Proximal_Promoter_df_filtered, Core_Promoter_df_filtered, Downstream_df_filtered\n",
    "# copynumber_filtered, gene_expression_filtered\n",
    "# survival, protein, cellsub, dense\n",
    "\n",
    "# Extract column names starting with 'TCGA' from methylation datasets\n",
    "tcga_columns_upstream = [col for col in Upstream_df_filtered.columns if col.startswith('TCGA')]\n",
    "tcga_columns_distal = [col for col in Distal_Promoter_df_filtered.columns if col.startswith('TCGA')]\n",
    "tcga_columns_proximal = [col for col in Proximal_Promoter_df_filtered.columns if col.startswith('TCGA')]\n",
    "tcga_columns_core = [col for col in Core_Promoter_df_filtered.columns if col.startswith('TCGA')]\n",
    "tcga_columns_downstream = [col for col in Downstream_df_filtered.columns if col.startswith('TCGA')]\n",
    "\n",
    "# Extract 'TCGA' columns from other datasets\n",
    "tcga_columns_copynumber = [col for col in copynumber_filtered.columns if col.startswith('TCGA')]\n",
    "tcga_columns_gene_expression = [col for col in gene_expression_filtered.columns if col.startswith('TCGA')]\n",
    "tcga_columns_survival = [col for col in survival['sample'] if col.startswith('TCGA')]\n",
    "tcga_columns_protein = [col for col in protein.columns if col.startswith('TCGA')]\n",
    "tcga_columns_cellsub = [col for col in cellsub['sampleID'] if col.startswith('TCGA')]\n",
    "tcga_columns_dense = [col for col in dense['sample'] if col.startswith('TCGA')]\n",
    "tcga_columns_immune_subtype = [col for col in immune_subtype['sample'] if col.startswith('TCGA')]\n",
    "# Find the intersection of TCGA column names across all DataFrames\n",
    "common_tcga_columns = set(tcga_columns_upstream) & set(tcga_columns_distal) & set(tcga_columns_proximal) & set(tcga_columns_core) & set(tcga_columns_downstream) & set(tcga_columns_copynumber) & set(tcga_columns_gene_expression) & set(tcga_columns_survival) & set(tcga_columns_protein) & set(tcga_columns_cellsub) & set(tcga_columns_dense) & set(tcga_columns_immune_subtype)\n",
    "\n",
    "# Convert the intersection back to a list, if needed\n",
    "common_tcga_columns_list = list(common_tcga_columns)\n",
    "\n",
    "# Print the number and the list of common TCGA columns\n",
    "print(f\"Number of common TCGA columns: {len(common_tcga_columns)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f32739c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define columns to keep along with common TCGA columns\n",
    "additional_cols_methylation = ['Closest_TSS_gene_name']\n",
    "\n",
    "# Filter each methylation DataFrame\n",
    "Upstream_df_filtered = Upstream_df_filtered[additional_cols_methylation + common_tcga_columns_list]\n",
    "Distal_Promoter_df_filtered = Distal_Promoter_df_filtered[additional_cols_methylation + common_tcga_columns_list]\n",
    "Proximal_Promoter_df_filtered = Proximal_Promoter_df_filtered[additional_cols_methylation + common_tcga_columns_list]\n",
    "Core_Promoter_df_filtered = Core_Promoter_df_filtered[additional_cols_methylation + common_tcga_columns_list]\n",
    "Downstream_df_filtered = Downstream_df_filtered[additional_cols_methylation + common_tcga_columns_list]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2502cfee",
   "metadata": {},
   "outputs": [],
   "source": [
    "Core_Promoter_df_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1513830b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#sort by gene name and reset index\n",
    "Upstream_df_filtered = Upstream_df_filtered.sort_values(by='Closest_TSS_gene_name').reset_index(drop=True)\n",
    "Distal_Promoter_df_filtered = Distal_Promoter_df_filtered.sort_values(by='Closest_TSS_gene_name').reset_index(drop=True)\n",
    "Proximal_Promoter_df_filtered = Proximal_Promoter_df_filtered.sort_values(by='Closest_TSS_gene_name').reset_index(drop=True)\n",
    "Core_Promoter_df_filtered = Core_Promoter_df_filtered.sort_values(by='Closest_TSS_gene_name').reset_index(drop=True)\n",
    "Downstream_df_filtered = Downstream_df_filtered.sort_values(by='Closest_TSS_gene_name').reset_index(drop=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "316b231a",
   "metadata": {},
   "outputs": [],
   "source": [
    "Upstream_df_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b095966",
   "metadata": {},
   "outputs": [],
   "source": [
    "copynumber_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0754561",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define columns to keep along with common TCGA columns\n",
    "additional_cols_copynumber = ['Sample']\n",
    "\n",
    "# Filter the copynumber DataFrame\n",
    "copynumber_filtered = copynumber_filtered[additional_cols_copynumber + common_tcga_columns_list]\n",
    "copynumber_filtered.columns.values[0] = 'gene'\n",
    "copynumber_filtered.sort_values(by='gene', inplace=True)\n",
    "copynumber_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e546d1e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define columns to keep along with common TCGA columns\n",
    "additional_cols_gene_expression = ['gene']\n",
    "\n",
    "# Filter the gene expression DataFrame\n",
    "gene_expression_filtered = gene_expression_filtered[additional_cols_gene_expression + common_tcga_columns_list]\n",
    "gene_expression_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3da470ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define columns to keep along with common TCGA columns\n",
    "additional_cols_protein = ['SampleID']\n",
    "\n",
    "# Filter the protein DataFrame\n",
    "protein_filtered = protein[additional_cols_protein + common_tcga_columns_list]\n",
    "protein_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3286a4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter rows based on common TCGA identifiers\n",
    "immune_subtype_filtered = immune_subtype[immune_subtype['sample'].isin(common_tcga_columns_list)]\n",
    "survival_filtered = survival[survival['sample'].isin(common_tcga_columns_list)]\n",
    "cellsub_filtered = cellsub[cellsub['sampleID'].isin(common_tcga_columns_list)]\n",
    "dense_filtered = dense[dense['sample'].isin(common_tcga_columns_list)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f23ca0ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "immune_subtype_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76db763c",
   "metadata": {},
   "outputs": [],
   "source": [
    "survival_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45601920",
   "metadata": {},
   "outputs": [],
   "source": [
    "cellsub_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82bccce0",
   "metadata": {},
   "outputs": [],
   "source": [
    "dense_filtered"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0952b86f",
   "metadata": {},
   "source": [
    "### 4.3 proteomics missing value and intersection\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98ba386e",
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cae9c59a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#calculate the NaN proportion of each row\n",
    "nan_proportions = protein_filtered.isna().mean(axis=1)\n",
    "# Display the results\n",
    "print(nan_proportions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01f4b486",
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_filtered = protein_filtered[nan_proportions <= 1/3]\n",
    "\n",
    "# Fill NaN values with 0 in the remaining rows\n",
    "protein_filtered = protein_filtered.fillna(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1f678bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_list = gene_expression_filtered['gene']\n",
    "protein_list = protein_filtered['SampleID'].tolist()\n",
    "protein_intersection = list(set(gene_list) & set(protein_list))\n",
    "len(protein_intersection)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d7123c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06cf726d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# select common genes in copynumber data\n",
    "protein_filtered = protein_filtered.loc[protein_filtered['SampleID'].isin(protein_intersection)].reset_index(drop=True)\n",
    "protein_filtered"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e11fa64",
   "metadata": {},
   "source": [
    "## 5. gene name/patient samples/ pheotype file lists"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cbd82019",
   "metadata": {},
   "source": [
    "### 5.1 gene name and patient samples lists"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2979e93",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_list = gene_expression_filtered['gene']\n",
    "gene_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3cfc08fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_list = protein_filtered['SampleID'].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(protein_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8ff7208",
   "metadata": {},
   "outputs": [],
   "source": [
    "intersection = list(set(gene_list) & set(protein_list))\n",
    "len(intersection)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31b4b22d",
   "metadata": {},
   "outputs": [],
   "source": [
    "patient_sample_list = pd.DataFrame(common_tcga_columns,columns=['sample'])\n",
    "patient_sample_list"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a84341da",
   "metadata": {},
   "source": [
    "### 5.2 phenotype lists"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0e77022",
   "metadata": {},
   "outputs": [],
   "source": [
    "immune_subtype_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7344f4be",
   "metadata": {},
   "outputs": [],
   "source": [
    "survival_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9cf6ccb",
   "metadata": {},
   "outputs": [],
   "source": [
    "survival_nan_column_proportions = survival_filtered.isna().mean()\n",
    "\n",
    "# Display the results\n",
    "print(survival_nan_column_proportions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b0475f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate the proportion of NaN values in each column\n",
    "survival_nan_column_proportions = survival_filtered.isna().mean()\n",
    "\n",
    "# Identify columns to be dropped (where proportion of NaN values is greater than 1/3)\n",
    "columns_to_drop = survival_nan_column_proportions[survival_nan_column_proportions > 1/3].index.tolist()\n",
    "\n",
    "# Drop these columns from the DataFrame\n",
    "survival_filtered = survival_filtered.drop(columns=columns_to_drop)\n",
    "\n",
    "# List of columns that were dropped\n",
    "print(\"Columns dropped:\", columns_to_drop)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42d0f9e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "survival_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99f9e454",
   "metadata": {},
   "outputs": [],
   "source": [
    "dense_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21bdeec1",
   "metadata": {},
   "outputs": [],
   "source": [
    "cellsub_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e41b34f",
   "metadata": {},
   "outputs": [],
   "source": [
    "cellsub_nan_column_proportions = cellsub_filtered.isna().mean()\n",
    "\n",
    "# Display the results\n",
    "print(cellsub_nan_column_proportions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c7af8f67",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate the proportion of NaN values in each column\n",
    "cellsub_nan_column_proportions = cellsub_filtered.isna().mean()\n",
    "\n",
    "# Identify columns to be dropped (where proportion of NaN values is greater than 1/3)\n",
    "columns_to_drop = cellsub_nan_column_proportions[cellsub_nan_column_proportions > 1/3].index.tolist()\n",
    "\n",
    "# Drop these columns from the DataFrame\n",
    "cellsub_filtered = cellsub_filtered.drop(columns=columns_to_drop)\n",
    "\n",
    "# List of columns that were dropped\n",
    "print(\"Columns dropped:\", columns_to_drop)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94a8b2d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "cellsub_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea31dd3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# extract phenotype names\n",
    "immune_phenotypes = immune_subtype_filtered.columns[1:].tolist()\n",
    "survival_phenotypes = survival_filtered.columns[2:].tolist() # _PATIENT infor is not needed (sample id)\n",
    "dense_phenotypes = dense_filtered.columns[2:].tolist() # sample_type_id infor is not needed (all = 1)\n",
    "cellsub_phenotypes = cellsub_filtered.columns[1:].tolist()\n",
    "\n",
    "# creat phenotype name and source\n",
    "phenotype_list = []\n",
    "phenotype_list.extend([(p, 'immunesub') for p in immune_phenotypes])\n",
    "phenotype_list.extend([(p, 'survival') for p in survival_phenotypes])\n",
    "phenotype_list.extend([(p, 'dense') for p in dense_phenotypes])\n",
    "phenotype_list.extend([(p, 'cellsub') for p in cellsub_phenotypes])\n",
    "\n",
    "# list DataFrame\n",
    "phenotype_lists = pd.DataFrame(phenotype_list, columns=['Phenotype_Name', 'Phenotype_Source'])\n",
    "phenotype_lists"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2bd57fbc",
   "metadata": {},
   "source": [
    "## 6. Save processed datasets"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 6.1 Keep the consistency for dataframes on genes and samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# [gene_list]\n",
    "# gene\n",
    "sorted_gene_list = gene_list.sort_values()\n",
    "sorted_gene = sorted_gene_list.tolist()\n",
    "sorted_gene_df = pd.DataFrame(sorted_gene, columns=['Gene'])\n",
    "display(sorted_gene_df)\n",
    "# gene-meth\n",
    "sorted_gene_methy = [gene + '-METH' for gene in sorted_gene]\n",
    "sorted_gene_methy_df = pd.DataFrame(sorted_gene_methy, columns=['Gene'])\n",
    "display(sorted_gene_methy_df)\n",
    "# gene-protein\n",
    "sorted_intersection = sorted(intersection)\n",
    "sorted_gene_protein = [gene + '-PROT' for gene in sorted_intersection]\n",
    "sorted_gene_protein_df = pd.DataFrame(sorted_gene_protein, columns=['Gene'])\n",
    "display(sorted_gene_protein_df)\n",
    "# all-gene\n",
    "sorted_gene_all = sorted_gene + sorted_gene_methy + sorted_gene_protein\n",
    "sorted_all_gene_df = pd.DataFrame(sorted_gene_all, columns=['Gene'])\n",
    "display(sorted_all_gene_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# [patient-sample-list]\n",
    "sorted_patient_sample_list = patient_sample_list.sort_values(by='sample')['sample'].tolist()\n",
    "print(sorted_patient_sample_list)\n",
    "sorted_patient_sample_df = patient_sample_list.sort_values(by='sample').reset_index(drop=True)\n",
    "display(sorted_patient_sample_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Upstream_df_filtered = Upstream_df_filtered[['Closest_TSS_gene_name'] + sorted_patient_sample_list]\n",
    "Upstream_df_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Distal_Promoter_df_filtered = Distal_Promoter_df_filtered[['Closest_TSS_gene_name'] + sorted_patient_sample_list]\n",
    "Distal_Promoter_df_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Proximal_Promoter_df_filtered = Proximal_Promoter_df_filtered[['Closest_TSS_gene_name'] + sorted_patient_sample_list]\n",
    "Proximal_Promoter_df_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Core_Promoter_df_filtered = Core_Promoter_df_filtered[['Closest_TSS_gene_name'] + sorted_patient_sample_list]\n",
    "Core_Promoter_df_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Downstream_df_filtered = Downstream_df_filtered[['Closest_TSS_gene_name'] + sorted_patient_sample_list]\n",
    "Downstream_df_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "copynumber_filtered = copynumber_filtered[['gene'] + sorted_patient_sample_list].sort_values(by='gene').reset_index(drop=True)\n",
    "copynumber_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_expression_filtered = gene_expression_filtered[['gene'] + sorted_patient_sample_list].sort_values(by='gene').reset_index(drop=True)\n",
    "gene_expression_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_filtered = protein_filtered[['SampleID'] + sorted_patient_sample_list].sort_values(by='SampleID').reset_index(drop=True)\n",
    "protein_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "immune_subtype_filtered = immune_subtype_filtered.sort_values(by='sample').reset_index(drop=True)\n",
    "immune_subtype_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "survival_filtered = survival_filtered.sort_values(by='sample').reset_index(drop=True)\n",
    "survival_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dense_filtered = dense_filtered.sort_values(by='sample').reset_index(drop=True)\n",
    "dense_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cellsub_filtered = cellsub_filtered.sort_values(by='sampleID').reset_index(drop=True)\n",
    "cellsub_filtered"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e63acd6",
   "metadata": {},
   "source": [
    "### 6.2 create output folder and save processed datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5512cfa",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "# outputfile name\n",
    "output_folder = 'UCSC-process'\n",
    "# create folder if not exist\n",
    "if not os.path.exists(output_folder):\n",
    "    os.makedirs(output_folder)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa76b540",
   "metadata": {},
   "outputs": [],
   "source": [
    "# DataFrame needed to be saved\n",
    "dataframes = {\n",
    "    'gene-list.csv': sorted_gene_df,\n",
    "    'gene-methy-list.csv': sorted_gene_methy_df,\n",
    "    'gene-protein-list.csv': sorted_gene_protein_df,\n",
    "    'gene-all-list.csv': sorted_all_gene_df,\n",
    "    'gene-kegg-edge-list.csv': filtered_up_kegg_df,\n",
    "    'gene-kegg-path-edge-list.csv': filtered_up_kegg_path_df,\n",
    "    'patient-sample-list.csv': sorted_patient_sample_df,\n",
    "    'phenotype-lists.csv': phenotype_lists,\n",
    "    'processed-genotype-methy-Upstream.csv': Upstream_df_filtered,\n",
    "    'processed-genotype-methy-Distal-Promoter.csv': Distal_Promoter_df_filtered,\n",
    "    'processed-genotype-methy-Proximal-Promoter.csv': Proximal_Promoter_df_filtered,\n",
    "    'processed-genotype-methy-Core-Promoter.csv': Core_Promoter_df_filtered,\n",
    "    'processed-genotype-methy-Downstream.csv': Downstream_df_filtered,\n",
    "    'processed-genotype-cnv.csv': copynumber_filtered,\n",
    "    'processed-genotype-gene-expression.csv': gene_expression_filtered,\n",
    "    'processed-genotype-proteomics.csv': protein_filtered,\n",
    "    'processed-phenotype-immune-subtype-transposed.csv': immune_subtype_filtered,\n",
    "    'processed-phenotype-survival-transposed.csv': survival_filtered,\n",
    "    'processed-phenotype-dense-transposed.csv': dense_filtered,\n",
    "    'processed-phenotype-cellsub-transposed.csv': cellsub_filtered\n",
    "}\n",
    "\n",
    "# save to output folder\n",
    "for file_name, df in dataframes.items():\n",
    "    df.to_csv(os.path.join(output_folder, file_name), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7. Convert the processed data into node dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load processed data\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "# read the file names under the folder\n",
    "# Define the path to the output folder where CSV files are stored\n",
    "output_folder = 'UCSC-process'\n",
    "\n",
    "# List of file names you saved earlier\n",
    "file_names = [\n",
    "    'gene-list', 'gene-methy-list', 'gene-protein-list', 'gene-all-list', \n",
    "    'gene-kegg-edge-list', 'gene-kegg-path-edge-list', 'patient-sample-list', \n",
    "    'phenotype-lists', 'processed-genotype-methy-Upstream', \n",
    "    'processed-genotype-methy-Distal-Promoter', \n",
    "    'processed-genotype-methy-Proximal-Promoter', \n",
    "    'processed-genotype-methy-Core-Promoter', 'processed-genotype-methy-Downstream', \n",
    "    'processed-genotype-cnv', 'processed-genotype-gene-expression', \n",
    "    'processed-genotype-proteomics', 'processed-phenotype-immune-subtype-transposed', \n",
    "    'processed-phenotype-survival-transposed', 'processed-phenotype-dense-transposed', \n",
    "    'processed-phenotype-cellsub-transposed'\n",
    "]\n",
    "\n",
    "# Dictionary to hold the dataframes\n",
    "dataframes = {}\n",
    "\n",
    "# Read each file and assign to a dataframe\n",
    "for file_name in file_names:\n",
    "    full_path = os.path.join(output_folder, file_name + '.csv')\n",
    "    dataframes[file_name] = pd.read_csv(full_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Assign each dataframe to a variable\n",
    "sorted_gene_df = dataframes['gene-list']\n",
    "sorted_gene_methy_df = dataframes['gene-methy-list']\n",
    "sorted_gene_protein_df = dataframes['gene-protein-list']\n",
    "sorted_all_gene_df = dataframes['gene-all-list']\n",
    "filtered_up_kegg_df = dataframes['gene-kegg-edge-list']\n",
    "filtered_up_kegg_path_df = dataframes['gene-kegg-path-edge-list']\n",
    "sorted_patient_sample_df = dataframes['patient-sample-list']\n",
    "phenotype_lists = dataframes['phenotype-lists']\n",
    "Upstream_df_filtered = dataframes['processed-genotype-methy-Upstream']\n",
    "Distal_Promoter_df_filtered = dataframes['processed-genotype-methy-Distal-Promoter']\n",
    "Proximal_Promoter_df_filtered = dataframes['processed-genotype-methy-Proximal-Promoter']\n",
    "Core_Promoter_df_filtered = dataframes['processed-genotype-methy-Core-Promoter']\n",
    "Downstream_df_filtered = dataframes['processed-genotype-methy-Downstream']\n",
    "copynumber_filtered = dataframes['processed-genotype-cnv']\n",
    "gene_expression_filtered = dataframes['processed-genotype-gene-expression']\n",
    "protein_filtered = dataframes['processed-genotype-proteomics']\n",
    "immune_subtype_filtered = dataframes['processed-phenotype-immune-subtype-transposed']\n",
    "survival_filtered = dataframes['processed-phenotype-survival-transposed']\n",
    "dense_filtered = dataframes['processed-phenotype-dense-transposed']\n",
    "cellsub_filtered = dataframes['processed-phenotype-cellsub-transposed']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# outputfile name\n",
    "graph_output_folder = 'graph-data'\n",
    "# create folder if not exist\n",
    "if not os.path.exists(graph_output_folder):\n",
    "    os.makedirs(graph_output_folder)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 7.1 Make nodes dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_num</th>\n",
       "      <th>Gene_name</th>\n",
       "      <th>NodeType</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>ABL1</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ABL2</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>ACAA1</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>ACACA</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>ACACB</td>\n",
       "      <td>Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4154</th>\n",
       "      <td>4154</td>\n",
       "      <td>SMAD4-PROT</td>\n",
       "      <td>Gene-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4155</th>\n",
       "      <td>4155</td>\n",
       "      <td>SRC-PROT</td>\n",
       "      <td>Gene-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4156</th>\n",
       "      <td>4156</td>\n",
       "      <td>SYK-PROT</td>\n",
       "      <td>Gene-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4157</th>\n",
       "      <td>4157</td>\n",
       "      <td>TFRC-PROT</td>\n",
       "      <td>Gene-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4158</th>\n",
       "      <td>4158</td>\n",
       "      <td>TSC1-PROT</td>\n",
       "      <td>Gene-PROT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4159 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Gene_num   Gene_name   NodeType\n",
       "0            0        ABL1       Gene\n",
       "1            1        ABL2       Gene\n",
       "2            2       ACAA1       Gene\n",
       "3            3       ACACA       Gene\n",
       "4            4       ACACB       Gene\n",
       "...        ...         ...        ...\n",
       "4154      4154  SMAD4-PROT  Gene-PROT\n",
       "4155      4155    SRC-PROT  Gene-PROT\n",
       "4156      4156    SYK-PROT  Gene-PROT\n",
       "4157      4157   TFRC-PROT  Gene-PROT\n",
       "4158      4158   TSC1-PROT  Gene-PROT\n",
       "\n",
       "[4159 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sorted_all_gene_dict = sorted_all_gene_df['Gene'].to_dict()\n",
    "sorted_all_gene_name_dict = {value: key for key, value in sorted_all_gene_dict.items()}\n",
    "num_gene = sorted_gene_df.shape[0]\n",
    "num_gene_protein = sorted_gene_protein_df.shape[0]\n",
    "nodetype_list = ['Gene'] * num_gene + ['Gene-METH'] * num_gene + ['Gene-PROT'] * num_gene_protein\n",
    "map_all_gene_df = pd.DataFrame({'Gene_num': sorted_all_gene_dict.keys(), 'Gene_name': sorted_all_gene_dict.values(), 'NodeType': nodetype_list})\n",
    "display(map_all_gene_df)\n",
    "map_all_gene_df.to_csv(os.path.join(graph_output_folder, 'map-all-gene.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 7.2 Create the edges connection between promoter methylations and proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>src</th>\n",
       "      <th>dest</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABL1-METH</td>\n",
       "      <td>ABL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABL2-METH</td>\n",
       "      <td>ABL2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACAA1-METH</td>\n",
       "      <td>ACAA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACACA-METH</td>\n",
       "      <td>ACACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACACB-METH</td>\n",
       "      <td>ACACB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2056</th>\n",
       "      <td>ZFYVE16-METH</td>\n",
       "      <td>ZFYVE16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2057</th>\n",
       "      <td>ZFYVE9-METH</td>\n",
       "      <td>ZFYVE9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2058</th>\n",
       "      <td>ZMAT3-METH</td>\n",
       "      <td>ZMAT3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2059</th>\n",
       "      <td>ZNF274-METH</td>\n",
       "      <td>ZNF274</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2060</th>\n",
       "      <td>ZNRF3-METH</td>\n",
       "      <td>ZNRF3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2061 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               src     dest\n",
       "0        ABL1-METH     ABL1\n",
       "1        ABL2-METH     ABL2\n",
       "2       ACAA1-METH    ACAA1\n",
       "3       ACACA-METH    ACACA\n",
       "4       ACACB-METH    ACACB\n",
       "...            ...      ...\n",
       "2056  ZFYVE16-METH  ZFYVE16\n",
       "2057   ZFYVE9-METH   ZFYVE9\n",
       "2058    ZMAT3-METH    ZMAT3\n",
       "2059   ZNF274-METH   ZNF274\n",
       "2060    ZNRF3-METH    ZNRF3\n",
       "\n",
       "[2061 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>src</th>\n",
       "      <th>dest</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ARAF</td>\n",
       "      <td>ARAF-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ATM</td>\n",
       "      <td>ATM-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BAX</td>\n",
       "      <td>BAX-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BCL2</td>\n",
       "      <td>BCL2-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BCL2A1</td>\n",
       "      <td>BCL2A1-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>BID</td>\n",
       "      <td>BID-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>BRAF</td>\n",
       "      <td>BRAF-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CDK1</td>\n",
       "      <td>CDK1-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>DUSP4</td>\n",
       "      <td>DUSP4-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>DVL3</td>\n",
       "      <td>DVL3-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>EEF2</td>\n",
       "      <td>EEF2-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>EEF2K</td>\n",
       "      <td>EEF2K-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>EGFR</td>\n",
       "      <td>EGFR-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>EIF4E</td>\n",
       "      <td>EIF4E-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ETS1</td>\n",
       "      <td>ETS1-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>FASN</td>\n",
       "      <td>FASN-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>GAB2</td>\n",
       "      <td>GAB2-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>GAPDH</td>\n",
       "      <td>GAPDH-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>IRF1</td>\n",
       "      <td>IRF1-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>IRS1</td>\n",
       "      <td>IRS1-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>JAK2</td>\n",
       "      <td>JAK2-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>LCK</td>\n",
       "      <td>LCK-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>MTOR</td>\n",
       "      <td>MTOR-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>NF2</td>\n",
       "      <td>NF2-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>NOTCH1</td>\n",
       "      <td>NOTCH1-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>NRAS</td>\n",
       "      <td>NRAS-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>PDK1</td>\n",
       "      <td>PDK1-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>PREX1</td>\n",
       "      <td>PREX1-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>PTEN</td>\n",
       "      <td>PTEN-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>RICTOR</td>\n",
       "      <td>RICTOR-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>SMAD1</td>\n",
       "      <td>SMAD1-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>SMAD3</td>\n",
       "      <td>SMAD3-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>SMAD4</td>\n",
       "      <td>SMAD4-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>SRC</td>\n",
       "      <td>SRC-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>SYK</td>\n",
       "      <td>SYK-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>TFRC</td>\n",
       "      <td>TFRC-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>TSC1</td>\n",
       "      <td>TSC1-PROT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       src         dest\n",
       "0     ARAF    ARAF-PROT\n",
       "1      ATM     ATM-PROT\n",
       "2      BAX     BAX-PROT\n",
       "3     BCL2    BCL2-PROT\n",
       "4   BCL2A1  BCL2A1-PROT\n",
       "5      BID     BID-PROT\n",
       "6     BRAF    BRAF-PROT\n",
       "7     CDK1    CDK1-PROT\n",
       "8    DUSP4   DUSP4-PROT\n",
       "9     DVL3    DVL3-PROT\n",
       "10    EEF2    EEF2-PROT\n",
       "11   EEF2K   EEF2K-PROT\n",
       "12    EGFR    EGFR-PROT\n",
       "13   EIF4E   EIF4E-PROT\n",
       "14    ETS1    ETS1-PROT\n",
       "15    FASN    FASN-PROT\n",
       "16    GAB2    GAB2-PROT\n",
       "17   GAPDH   GAPDH-PROT\n",
       "18    IRF1    IRF1-PROT\n",
       "19    IRS1    IRS1-PROT\n",
       "20    JAK2    JAK2-PROT\n",
       "21     LCK     LCK-PROT\n",
       "22    MTOR    MTOR-PROT\n",
       "23     NF2     NF2-PROT\n",
       "24  NOTCH1  NOTCH1-PROT\n",
       "25    NRAS    NRAS-PROT\n",
       "26    PDK1    PDK1-PROT\n",
       "27   PREX1   PREX1-PROT\n",
       "28    PTEN    PTEN-PROT\n",
       "29  RICTOR  RICTOR-PROT\n",
       "30   SMAD1   SMAD1-PROT\n",
       "31   SMAD3   SMAD3-PROT\n",
       "32   SMAD4   SMAD4-PROT\n",
       "33     SRC     SRC-PROT\n",
       "34     SYK     SYK-PROT\n",
       "35    TFRC    TFRC-PROT\n",
       "36    TSC1    TSC1-PROT"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# [Gene-METH - Gene]\n",
    "sorted_gene_methy = sorted_gene_methy_df['Gene'].tolist()\n",
    "sorted_gene_list = sorted_gene_df['Gene'].tolist()\n",
    "sorted_gene_protein = sorted_gene_protein_df['Gene'].tolist()\n",
    "sorted_intersection = [gene_protein.replace('-PROT', '') for gene_protein in sorted_gene_protein]\n",
    "gene_meth_edge_df = pd.DataFrame({'src': sorted_gene_methy, 'dest': sorted_gene_list})\n",
    "display(gene_meth_edge_df)\n",
    "# [Gene - Gene-PROT]\n",
    "gene_protein_edge_df = pd.DataFrame({'src': sorted_intersection, 'dest': sorted_gene_protein})\n",
    "display(gene_protein_edge_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2061"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sorted_all_gene_name_dict['ABL1-METH']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>src</th>\n",
       "      <th>dest</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2061</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2062</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2063</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2064</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2065</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2056</th>\n",
       "      <td>4117</td>\n",
       "      <td>2056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2057</th>\n",
       "      <td>4118</td>\n",
       "      <td>2057</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2058</th>\n",
       "      <td>4119</td>\n",
       "      <td>2058</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2059</th>\n",
       "      <td>4120</td>\n",
       "      <td>2059</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2060</th>\n",
       "      <td>4121</td>\n",
       "      <td>2060</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2061 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       src  dest\n",
       "0     2061     0\n",
       "1     2062     1\n",
       "2     2063     2\n",
       "3     2064     3\n",
       "4     2065     4\n",
       "...    ...   ...\n",
       "2056  4117  2056\n",
       "2057  4118  2057\n",
       "2058  4119  2058\n",
       "2059  4120  2059\n",
       "2060  4121  2060\n",
       "\n",
       "[2061 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>src</th>\n",
       "      <th>dest</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>92</td>\n",
       "      <td>4122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>110</td>\n",
       "      <td>4123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>152</td>\n",
       "      <td>4124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>156</td>\n",
       "      <td>4125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>157</td>\n",
       "      <td>4126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>166</td>\n",
       "      <td>4127</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>183</td>\n",
       "      <td>4128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>331</td>\n",
       "      <td>4129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>525</td>\n",
       "      <td>4130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>533</td>\n",
       "      <td>4131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>547</td>\n",
       "      <td>4132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>548</td>\n",
       "      <td>4133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>555</td>\n",
       "      <td>4134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>565</td>\n",
       "      <td>4135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>589</td>\n",
       "      <td>4136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>608</td>\n",
       "      <td>4137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>687</td>\n",
       "      <td>4138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>697</td>\n",
       "      <td>4139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>957</td>\n",
       "      <td>4140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>962</td>\n",
       "      <td>4141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1000</td>\n",
       "      <td>4142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>1080</td>\n",
       "      <td>4143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>1244</td>\n",
       "      <td>4144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1275</td>\n",
       "      <td>4145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1304</td>\n",
       "      <td>4146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1322</td>\n",
       "      <td>4147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>1388</td>\n",
       "      <td>4148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>1535</td>\n",
       "      <td>4149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>1579</td>\n",
       "      <td>4150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>1674</td>\n",
       "      <td>4151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>1799</td>\n",
       "      <td>4152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>1801</td>\n",
       "      <td>4153</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>1802</td>\n",
       "      <td>4154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>1836</td>\n",
       "      <td>4155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>1864</td>\n",
       "      <td>4156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>1902</td>\n",
       "      <td>4157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>1984</td>\n",
       "      <td>4158</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     src  dest\n",
       "0     92  4122\n",
       "1    110  4123\n",
       "2    152  4124\n",
       "3    156  4125\n",
       "4    157  4126\n",
       "5    166  4127\n",
       "6    183  4128\n",
       "7    331  4129\n",
       "8    525  4130\n",
       "9    533  4131\n",
       "10   547  4132\n",
       "11   548  4133\n",
       "12   555  4134\n",
       "13   565  4135\n",
       "14   589  4136\n",
       "15   608  4137\n",
       "16   687  4138\n",
       "17   697  4139\n",
       "18   957  4140\n",
       "19   962  4141\n",
       "20  1000  4142\n",
       "21  1080  4143\n",
       "22  1244  4144\n",
       "23  1275  4145\n",
       "24  1304  4146\n",
       "25  1322  4147\n",
       "26  1388  4148\n",
       "27  1535  4149\n",
       "28  1579  4150\n",
       "29  1674  4151\n",
       "30  1799  4152\n",
       "31  1801  4153\n",
       "32  1802  4154\n",
       "33  1836  4155\n",
       "34  1864  4156\n",
       "35  1902  4157\n",
       "36  1984  4158"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# replace gene name with gene number\n",
    "gene_meth_num_edge_df = gene_meth_edge_df.copy()\n",
    "gene_meth_num_edge_df['src'] = gene_meth_edge_df['src'].map(sorted_all_gene_name_dict)\n",
    "gene_meth_num_edge_df['dest'] = gene_meth_edge_df['dest'].map(sorted_all_gene_name_dict)\n",
    "display(gene_meth_num_edge_df)\n",
    "gene_protein_num_edge_df = gene_protein_edge_df.copy()\n",
    "gene_protein_num_edge_df['src'] = gene_protein_edge_df['src'].map(sorted_all_gene_name_dict)\n",
    "gene_protein_num_edge_df['dest'] = gene_protein_edge_df['dest'].map(sorted_all_gene_name_dict)\n",
    "display(gene_protein_num_edge_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 7.3 Concat all of the edges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>src</th>\n",
       "      <th>dest</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>18</td>\n",
       "      <td>1255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>18</td>\n",
       "      <td>1256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>18</td>\n",
       "      <td>1257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>18</td>\n",
       "      <td>1857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>19</td>\n",
       "      <td>1255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18226</th>\n",
       "      <td>2060</td>\n",
       "      <td>679</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18227</th>\n",
       "      <td>2060</td>\n",
       "      <td>680</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18228</th>\n",
       "      <td>2060</td>\n",
       "      <td>681</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18229</th>\n",
       "      <td>2060</td>\n",
       "      <td>682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18230</th>\n",
       "      <td>2060</td>\n",
       "      <td>683</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18231 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        src  dest\n",
       "0        18  1255\n",
       "1        18  1256\n",
       "2        18  1257\n",
       "3        18  1857\n",
       "4        19  1255\n",
       "...     ...   ...\n",
       "18226  2060   679\n",
       "18227  2060   680\n",
       "18228  2060   681\n",
       "18229  2060   682\n",
       "18230  2060   683\n",
       "\n",
       "[18231 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>src</th>\n",
       "      <th>dest</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>18</td>\n",
       "      <td>1255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>18</td>\n",
       "      <td>1256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>18</td>\n",
       "      <td>1257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>18</td>\n",
       "      <td>1857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>19</td>\n",
       "      <td>1255</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>1802</td>\n",
       "      <td>4154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>1836</td>\n",
       "      <td>4155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>1864</td>\n",
       "      <td>4156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>1902</td>\n",
       "      <td>4157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>1984</td>\n",
       "      <td>4158</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20329 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     src  dest\n",
       "0     18  1255\n",
       "1     18  1256\n",
       "2     18  1257\n",
       "3     18  1857\n",
       "4     19  1255\n",
       "..   ...   ...\n",
       "32  1802  4154\n",
       "33  1836  4155\n",
       "34  1864  4156\n",
       "35  1902  4157\n",
       "36  1984  4158\n",
       "\n",
       "[20329 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>src</th>\n",
       "      <th>dest</th>\n",
       "      <th>EdgeType</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>18</td>\n",
       "      <td>1255</td>\n",
       "      <td>Gene-Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>18</td>\n",
       "      <td>1256</td>\n",
       "      <td>Gene-Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>18</td>\n",
       "      <td>1257</td>\n",
       "      <td>Gene-Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>18</td>\n",
       "      <td>1857</td>\n",
       "      <td>Gene-Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>19</td>\n",
       "      <td>1255</td>\n",
       "      <td>Gene-Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>1802</td>\n",
       "      <td>4154</td>\n",
       "      <td>Gene-Gene-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>1836</td>\n",
       "      <td>4155</td>\n",
       "      <td>Gene-Gene-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>1864</td>\n",
       "      <td>4156</td>\n",
       "      <td>Gene-Gene-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>1902</td>\n",
       "      <td>4157</td>\n",
       "      <td>Gene-Gene-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>1984</td>\n",
       "      <td>4158</td>\n",
       "      <td>Gene-Gene-PROT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20329 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     src  dest        EdgeType\n",
       "0     18  1255       Gene-Gene\n",
       "1     18  1256       Gene-Gene\n",
       "2     18  1257       Gene-Gene\n",
       "3     18  1857       Gene-Gene\n",
       "4     19  1255       Gene-Gene\n",
       "..   ...   ...             ...\n",
       "32  1802  4154  Gene-Gene-PROT\n",
       "33  1836  4155  Gene-Gene-PROT\n",
       "34  1864  4156  Gene-Gene-PROT\n",
       "35  1902  4157  Gene-Gene-PROT\n",
       "36  1984  4158  Gene-Gene-PROT\n",
       "\n",
       "[20329 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "filtered_up_kegg_num_df = filtered_up_kegg_df.copy()\n",
    "filtered_up_kegg_num_df['src'] = filtered_up_kegg_num_df['src'].map(sorted_all_gene_name_dict)\n",
    "filtered_up_kegg_num_df['dest'] = filtered_up_kegg_num_df['dest'].map(sorted_all_gene_name_dict)\n",
    "display(filtered_up_kegg_num_df)\n",
    "all_gene_edge_num_df = pd.concat([filtered_up_kegg_num_df, gene_meth_num_edge_df, gene_protein_num_edge_df])\n",
    "display(all_gene_edge_num_df)\n",
    "\n",
    "num_gene_edge = filtered_up_kegg_num_df.shape[0]\n",
    "num_gene_meth_edge = gene_meth_num_edge_df.shape[0]\n",
    "num_gene_protein_edge = gene_protein_num_edge_df.shape[0]\n",
    "edgetype_list = ['Gene-Gene'] * num_gene_edge + ['Gene-Gene-METH'] * num_gene_meth_edge + ['Gene-Gene-PROT'] * num_gene_protein_edge\n",
    "all_gene_edge_num_df['EdgeType'] = edgetype_list\n",
    "all_gene_edge_num_df = all_gene_edge_num_df.sort_values(by=['src', 'dest']).reset_index(drop=True)\n",
    "display(all_gene_edge_num_df)\n",
    "all_gene_edge_num_df.to_csv(os.path.join(graph_output_folder, 'all-gene-edge-num.csv'), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>src</th>\n",
       "      <th>dest</th>\n",
       "      <th>EdgeType</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACTB</td>\n",
       "      <td>MYL6</td>\n",
       "      <td>Gene-Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACTB</td>\n",
       "      <td>MYL6B</td>\n",
       "      <td>Gene-Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACTB</td>\n",
       "      <td>MYL9</td>\n",
       "      <td>Gene-Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ACTB</td>\n",
       "      <td>STK3</td>\n",
       "      <td>Gene-Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACTG1</td>\n",
       "      <td>MYL6</td>\n",
       "      <td>Gene-Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>SMAD4</td>\n",
       "      <td>SMAD4-PROT</td>\n",
       "      <td>Gene-Gene-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>SRC</td>\n",
       "      <td>SRC-PROT</td>\n",
       "      <td>Gene-Gene-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>SYK</td>\n",
       "      <td>SYK-PROT</td>\n",
       "      <td>Gene-Gene-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>TFRC</td>\n",
       "      <td>TFRC-PROT</td>\n",
       "      <td>Gene-Gene-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>TSC1</td>\n",
       "      <td>TSC1-PROT</td>\n",
       "      <td>Gene-Gene-PROT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20329 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      src        dest        EdgeType\n",
       "0    ACTB        MYL6       Gene-Gene\n",
       "1    ACTB       MYL6B       Gene-Gene\n",
       "2    ACTB        MYL9       Gene-Gene\n",
       "3    ACTB        STK3       Gene-Gene\n",
       "4   ACTG1        MYL6       Gene-Gene\n",
       "..    ...         ...             ...\n",
       "32  SMAD4  SMAD4-PROT  Gene-Gene-PROT\n",
       "33    SRC    SRC-PROT  Gene-Gene-PROT\n",
       "34    SYK    SYK-PROT  Gene-Gene-PROT\n",
       "35   TFRC   TFRC-PROT  Gene-Gene-PROT\n",
       "36   TSC1   TSC1-PROT  Gene-Gene-PROT\n",
       "\n",
       "[20329 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# gene edge interactions without map\n",
    "all_gene_edge_df = all_gene_edge_num_df.copy()\n",
    "all_gene_edge_df = all_gene_edge_df.replace(sorted_all_gene_dict)\n",
    "\n",
    "num_gene_edge = filtered_up_kegg_num_df.shape[0]\n",
    "num_gene_meth_edge = gene_meth_edge_df.shape[0]\n",
    "num_gene_protein_edge = gene_protein_edge_df.shape[0]\n",
    "# all_gene_edge_df = all_gene_edge_df.sort_values(by=['src', 'dest']).reset_index(drop=True)\n",
    "all_gene_edge_df.to_csv(os.path.join(graph_output_folder, 'all-gene-edge.csv'), index=False)\n",
    "display(all_gene_edge_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 8. Load data into graph format"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 8.1 Form up the input samples"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "recommends the use of the endpoints of OS, PFI, DFI, and DSS for each TCGA cancer type\n",
    "\n",
    "* OS: overall survial\n",
    "* PFI: progression-free interval\n",
    "* DSS: disease-specific survival\n",
    "* DFI: disease-free interval"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>_PATIENT</th>\n",
       "      <th>cancer type abbreviation</th>\n",
       "      <th>age_at_initial_pathologic_diagnosis</th>\n",
       "      <th>gender</th>\n",
       "      <th>race</th>\n",
       "      <th>ajcc_pathologic_tumor_stage</th>\n",
       "      <th>histological_type</th>\n",
       "      <th>initial_pathologic_dx_year</th>\n",
       "      <th>birth_days_to</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>tumor_status</th>\n",
       "      <th>last_contact_days_to</th>\n",
       "      <th>OS</th>\n",
       "      <th>OS.time</th>\n",
       "      <th>DSS</th>\n",
       "      <th>DSS.time</th>\n",
       "      <th>PFI</th>\n",
       "      <th>PFI.time</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-05-4384-01</td>\n",
       "      <td>TCGA-05-4384</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>66.0</td>\n",
       "      <td>MALE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Stage IIIA</td>\n",
       "      <td>Lung Adenocarcinoma</td>\n",
       "      <td>2009.0</td>\n",
       "      <td>-24411.0</td>\n",
       "      <td>Alive</td>\n",
       "      <td>WITH TUMOR</td>\n",
       "      <td>426.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>426.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>426.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>183.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-05-4396-01</td>\n",
       "      <td>TCGA-05-4396</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>76.0</td>\n",
       "      <td>MALE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Stage IIIB</td>\n",
       "      <td>Lung Adenocarcinoma</td>\n",
       "      <td>2006.0</td>\n",
       "      <td>-28094.0</td>\n",
       "      <td>Dead</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>303.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>303.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>303.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-05-4405-01</td>\n",
       "      <td>TCGA-05-4405</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>74.0</td>\n",
       "      <td>FEMALE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Stage IB</td>\n",
       "      <td>Lung Adenocarcinoma</td>\n",
       "      <td>2006.0</td>\n",
       "      <td>-27241.0</td>\n",
       "      <td>Alive</td>\n",
       "      <td>TUMOR FREE</td>\n",
       "      <td>610.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>610.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>610.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>610.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-05-4410-01</td>\n",
       "      <td>TCGA-05-4410</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>62.0</td>\n",
       "      <td>MALE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Stage IB</td>\n",
       "      <td>Lung Adenocarcinoma</td>\n",
       "      <td>2007.0</td>\n",
       "      <td>-22888.0</td>\n",
       "      <td>Alive</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-05-4417-01</td>\n",
       "      <td>TCGA-05-4417</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>51.0</td>\n",
       "      <td>FEMALE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Stage IB</td>\n",
       "      <td>Lung Adenocarcinoma</td>\n",
       "      <td>2008.0</td>\n",
       "      <td>-18780.0</td>\n",
       "      <td>Alive</td>\n",
       "      <td>TUMOR FREE</td>\n",
       "      <td>455.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>455.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>455.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>455.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3657</th>\n",
       "      <td>TCGA-ZA-A8F6-01</td>\n",
       "      <td>TCGA-ZA-A8F6</td>\n",
       "      <td>STAD</td>\n",
       "      <td>71.0</td>\n",
       "      <td>MALE</td>\n",
       "      <td>WHITE</td>\n",
       "      <td>Stage IB</td>\n",
       "      <td>Stomach, Intestinal Adenocarcinoma, Not Otherw...</td>\n",
       "      <td>2013.0</td>\n",
       "      <td>-26122.0</td>\n",
       "      <td>Alive</td>\n",
       "      <td>TUMOR FREE</td>\n",
       "      <td>525.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>525.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>525.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>525.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3658</th>\n",
       "      <td>TCGA-ZG-A8QW-01</td>\n",
       "      <td>TCGA-ZG-A8QW</td>\n",
       "      <td>PRAD</td>\n",
       "      <td>72.0</td>\n",
       "      <td>MALE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Prostate Adenocarcinoma Acinar Type</td>\n",
       "      <td>2013.0</td>\n",
       "      <td>-26632.0</td>\n",
       "      <td>Alive</td>\n",
       "      <td>TUMOR FREE</td>\n",
       "      <td>94.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>94.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>94.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>94.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3659</th>\n",
       "      <td>TCGA-ZG-A8QX-01</td>\n",
       "      <td>TCGA-ZG-A8QX</td>\n",
       "      <td>PRAD</td>\n",
       "      <td>56.0</td>\n",
       "      <td>MALE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Prostate Adenocarcinoma Acinar Type</td>\n",
       "      <td>2013.0</td>\n",
       "      <td>-20683.0</td>\n",
       "      <td>Alive</td>\n",
       "      <td>TUMOR FREE</td>\n",
       "      <td>442.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>442.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>442.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>442.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3660</th>\n",
       "      <td>TCGA-ZG-A8QY-01</td>\n",
       "      <td>TCGA-ZG-A8QY</td>\n",
       "      <td>PRAD</td>\n",
       "      <td>67.0</td>\n",
       "      <td>MALE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Prostate Adenocarcinoma Acinar Type</td>\n",
       "      <td>2013.0</td>\n",
       "      <td>-24731.0</td>\n",
       "      <td>Alive</td>\n",
       "      <td>WITH TUMOR</td>\n",
       "      <td>404.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>404.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>404.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>404.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3661</th>\n",
       "      <td>TCGA-ZG-A8QZ-01</td>\n",
       "      <td>TCGA-ZG-A8QZ</td>\n",
       "      <td>PRAD</td>\n",
       "      <td>65.0</td>\n",
       "      <td>MALE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Prostate Adenocarcinoma Acinar Type</td>\n",
       "      <td>2013.0</td>\n",
       "      <td>-23980.0</td>\n",
       "      <td>Alive</td>\n",
       "      <td>TUMOR FREE</td>\n",
       "      <td>308.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>308.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>308.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>308.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3662 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               sample      _PATIENT cancer type abbreviation  \\\n",
       "0     TCGA-05-4384-01  TCGA-05-4384                     LUAD   \n",
       "1     TCGA-05-4396-01  TCGA-05-4396                     LUAD   \n",
       "2     TCGA-05-4405-01  TCGA-05-4405                     LUAD   \n",
       "3     TCGA-05-4410-01  TCGA-05-4410                     LUAD   \n",
       "4     TCGA-05-4417-01  TCGA-05-4417                     LUAD   \n",
       "...               ...           ...                      ...   \n",
       "3657  TCGA-ZA-A8F6-01  TCGA-ZA-A8F6                     STAD   \n",
       "3658  TCGA-ZG-A8QW-01  TCGA-ZG-A8QW                     PRAD   \n",
       "3659  TCGA-ZG-A8QX-01  TCGA-ZG-A8QX                     PRAD   \n",
       "3660  TCGA-ZG-A8QY-01  TCGA-ZG-A8QY                     PRAD   \n",
       "3661  TCGA-ZG-A8QZ-01  TCGA-ZG-A8QZ                     PRAD   \n",
       "\n",
       "      age_at_initial_pathologic_diagnosis  gender   race  \\\n",
       "0                                    66.0    MALE    NaN   \n",
       "1                                    76.0    MALE    NaN   \n",
       "2                                    74.0  FEMALE    NaN   \n",
       "3                                    62.0    MALE    NaN   \n",
       "4                                    51.0  FEMALE    NaN   \n",
       "...                                   ...     ...    ...   \n",
       "3657                                 71.0    MALE  WHITE   \n",
       "3658                                 72.0    MALE    NaN   \n",
       "3659                                 56.0    MALE    NaN   \n",
       "3660                                 67.0    MALE    NaN   \n",
       "3661                                 65.0    MALE    NaN   \n",
       "\n",
       "     ajcc_pathologic_tumor_stage  \\\n",
       "0                     Stage IIIA   \n",
       "1                     Stage IIIB   \n",
       "2                       Stage IB   \n",
       "3                       Stage IB   \n",
       "4                       Stage IB   \n",
       "...                          ...   \n",
       "3657                    Stage IB   \n",
       "3658                         NaN   \n",
       "3659                         NaN   \n",
       "3660                         NaN   \n",
       "3661                         NaN   \n",
       "\n",
       "                                      histological_type  \\\n",
       "0                                   Lung Adenocarcinoma   \n",
       "1                                   Lung Adenocarcinoma   \n",
       "2                                   Lung Adenocarcinoma   \n",
       "3                                   Lung Adenocarcinoma   \n",
       "4                                   Lung Adenocarcinoma   \n",
       "...                                                 ...   \n",
       "3657  Stomach, Intestinal Adenocarcinoma, Not Otherw...   \n",
       "3658                Prostate Adenocarcinoma Acinar Type   \n",
       "3659                Prostate Adenocarcinoma Acinar Type   \n",
       "3660                Prostate Adenocarcinoma Acinar Type   \n",
       "3661                Prostate Adenocarcinoma Acinar Type   \n",
       "\n",
       "      initial_pathologic_dx_year  birth_days_to vital_status tumor_status  \\\n",
       "0                         2009.0       -24411.0        Alive   WITH TUMOR   \n",
       "1                         2006.0       -28094.0         Dead          NaN   \n",
       "2                         2006.0       -27241.0        Alive   TUMOR FREE   \n",
       "3                         2007.0       -22888.0        Alive          NaN   \n",
       "4                         2008.0       -18780.0        Alive   TUMOR FREE   \n",
       "...                          ...            ...          ...          ...   \n",
       "3657                      2013.0       -26122.0        Alive   TUMOR FREE   \n",
       "3658                      2013.0       -26632.0        Alive   TUMOR FREE   \n",
       "3659                      2013.0       -20683.0        Alive   TUMOR FREE   \n",
       "3660                      2013.0       -24731.0        Alive   WITH TUMOR   \n",
       "3661                      2013.0       -23980.0        Alive   TUMOR FREE   \n",
       "\n",
       "      last_contact_days_to   OS  OS.time  DSS  DSS.time  PFI  PFI.time  \n",
       "0                    426.0  0.0    426.0  0.0     426.0  1.0     183.0  \n",
       "1                      NaN  1.0    303.0  NaN     303.0  0.0     303.0  \n",
       "2                    610.0  0.0    610.0  0.0     610.0  0.0     610.0  \n",
       "3                      0.0  0.0      0.0  0.0       0.0  0.0       0.0  \n",
       "4                    455.0  0.0    455.0  0.0     455.0  0.0     455.0  \n",
       "...                    ...  ...      ...  ...       ...  ...       ...  \n",
       "3657                 525.0  0.0    525.0  0.0     525.0  0.0     525.0  \n",
       "3658                  94.0  0.0     94.0  0.0      94.0  0.0      94.0  \n",
       "3659                 442.0  0.0    442.0  0.0     442.0  0.0     442.0  \n",
       "3660                 404.0  0.0    404.0  0.0     404.0  0.0     404.0  \n",
       "3661                 308.0  0.0    308.0  0.0     308.0  0.0     308.0  \n",
       "\n",
       "[3662 rows x 19 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "survival_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>cancer type abbreviation</th>\n",
       "      <th>OS</th>\n",
       "      <th>vital_status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-05-4384-01</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Alive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-05-4396-01</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Dead</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-05-4405-01</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Alive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-05-4410-01</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Alive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-05-4417-01</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Alive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3657</th>\n",
       "      <td>TCGA-ZA-A8F6-01</td>\n",
       "      <td>STAD</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Alive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3658</th>\n",
       "      <td>TCGA-ZG-A8QW-01</td>\n",
       "      <td>PRAD</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Alive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3659</th>\n",
       "      <td>TCGA-ZG-A8QX-01</td>\n",
       "      <td>PRAD</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Alive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3660</th>\n",
       "      <td>TCGA-ZG-A8QY-01</td>\n",
       "      <td>PRAD</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Alive</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3661</th>\n",
       "      <td>TCGA-ZG-A8QZ-01</td>\n",
       "      <td>PRAD</td>\n",
       "      <td>0.0</td>\n",
       "      <td>Alive</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3662 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               sample cancer type abbreviation   OS vital_status\n",
       "0     TCGA-05-4384-01                     LUAD  0.0        Alive\n",
       "1     TCGA-05-4396-01                     LUAD  1.0         Dead\n",
       "2     TCGA-05-4405-01                     LUAD  0.0        Alive\n",
       "3     TCGA-05-4410-01                     LUAD  0.0        Alive\n",
       "4     TCGA-05-4417-01                     LUAD  0.0        Alive\n",
       "...               ...                      ...  ...          ...\n",
       "3657  TCGA-ZA-A8F6-01                     STAD  0.0        Alive\n",
       "3658  TCGA-ZG-A8QW-01                     PRAD  0.0        Alive\n",
       "3659  TCGA-ZG-A8QX-01                     PRAD  0.0        Alive\n",
       "3660  TCGA-ZG-A8QY-01                     PRAD  0.0        Alive\n",
       "3661  TCGA-ZG-A8QZ-01                     PRAD  0.0        Alive\n",
       "\n",
       "[3662 rows x 4 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sample                      0\n",
      "cancer type abbreviation    0\n",
      "OS                          0\n",
      "vital_status                0\n",
      "dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>cancer type abbreviation</th>\n",
       "      <th>OS</th>\n",
       "      <th>vital_status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-05-4384-01</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-05-4396-01</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-05-4405-01</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-05-4410-01</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-05-4417-01</td>\n",
       "      <td>LUAD</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3657</th>\n",
       "      <td>TCGA-ZA-A8F6-01</td>\n",
       "      <td>STAD</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3658</th>\n",
       "      <td>TCGA-ZG-A8QW-01</td>\n",
       "      <td>PRAD</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3659</th>\n",
       "      <td>TCGA-ZG-A8QX-01</td>\n",
       "      <td>PRAD</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3660</th>\n",
       "      <td>TCGA-ZG-A8QY-01</td>\n",
       "      <td>PRAD</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3661</th>\n",
       "      <td>TCGA-ZG-A8QZ-01</td>\n",
       "      <td>PRAD</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3662 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               sample cancer type abbreviation   OS  vital_status\n",
       "0     TCGA-05-4384-01                     LUAD  0.0           0.0\n",
       "1     TCGA-05-4396-01                     LUAD  1.0           1.0\n",
       "2     TCGA-05-4405-01                     LUAD  0.0           0.0\n",
       "3     TCGA-05-4410-01                     LUAD  0.0           0.0\n",
       "4     TCGA-05-4417-01                     LUAD  0.0           0.0\n",
       "...               ...                      ...  ...           ...\n",
       "3657  TCGA-ZA-A8F6-01                     STAD  0.0           0.0\n",
       "3658  TCGA-ZG-A8QW-01                     PRAD  0.0           0.0\n",
       "3659  TCGA-ZG-A8QX-01                     PRAD  0.0           0.0\n",
       "3660  TCGA-ZG-A8QY-01                     PRAD  0.0           0.0\n",
       "3661  TCGA-ZG-A8QZ-01                     PRAD  0.0           0.0\n",
       "\n",
       "[3662 rows x 4 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "0       True\n",
       "1       True\n",
       "2       True\n",
       "3       True\n",
       "4       True\n",
       "        ... \n",
       "3657    True\n",
       "3658    True\n",
       "3659    True\n",
       "3660    True\n",
       "3661    True\n",
       "Length: 3662, dtype: bool"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "survival_filtered_feature_df = survival_filtered.copy()\n",
    "survival_filtered_feature_df = survival_filtered_feature_df[['sample', 'cancer type abbreviation', 'OS', 'vital_status']]\n",
    "display(survival_filtered_feature_df)\n",
    "\n",
    "nan_counts = survival_filtered_feature_df.isna().sum()  # or df.isnull()\n",
    "print(nan_counts)\n",
    "\n",
    "# Convert 'alive' to 0.0 and 'dead' to 1.0\n",
    "survival_filtered_feature_df['vital_status'] = survival_filtered_feature_df['vital_status'].map({'Alive': 0.0, 'Dead': 1.0})\n",
    "display(survival_filtered_feature_df)\n",
    "survival_filtered_feature_df['OS'] == survival_filtered_feature_df['vital_status']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All rows have the same value in column 'OS' and column 'vital_status' : True\n"
     ]
    }
   ],
   "source": [
    "# Check if each row in Column1 and Column2 have the same value\n",
    "rows_same = (survival_filtered_feature_df['OS'] == survival_filtered_feature_df['vital_status']).all()\n",
    "print(\"All rows have the same value in column 'OS' and column 'vital_status' :\", rows_same)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>OS</th>\n",
       "      <th>cancer type abbreviation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-05-4384-01</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LUAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-05-4396-01</td>\n",
       "      <td>1.0</td>\n",
       "      <td>LUAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-05-4405-01</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LUAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-05-4410-01</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LUAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-05-4417-01</td>\n",
       "      <td>0.0</td>\n",
       "      <td>LUAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3657</th>\n",
       "      <td>TCGA-ZA-A8F6-01</td>\n",
       "      <td>0.0</td>\n",
       "      <td>STAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3658</th>\n",
       "      <td>TCGA-ZG-A8QW-01</td>\n",
       "      <td>0.0</td>\n",
       "      <td>PRAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3659</th>\n",
       "      <td>TCGA-ZG-A8QX-01</td>\n",
       "      <td>0.0</td>\n",
       "      <td>PRAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3660</th>\n",
       "      <td>TCGA-ZG-A8QY-01</td>\n",
       "      <td>0.0</td>\n",
       "      <td>PRAD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3661</th>\n",
       "      <td>TCGA-ZG-A8QZ-01</td>\n",
       "      <td>0.0</td>\n",
       "      <td>PRAD</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3662 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               sample   OS cancer type abbreviation\n",
       "0     TCGA-05-4384-01  0.0                     LUAD\n",
       "1     TCGA-05-4396-01  1.0                     LUAD\n",
       "2     TCGA-05-4405-01  0.0                     LUAD\n",
       "3     TCGA-05-4410-01  0.0                     LUAD\n",
       "4     TCGA-05-4417-01  0.0                     LUAD\n",
       "...               ...  ...                      ...\n",
       "3657  TCGA-ZA-A8F6-01  0.0                     STAD\n",
       "3658  TCGA-ZG-A8QW-01  0.0                     PRAD\n",
       "3659  TCGA-ZG-A8QX-01  0.0                     PRAD\n",
       "3660  TCGA-ZG-A8QY-01  0.0                     PRAD\n",
       "3661  TCGA-ZG-A8QZ-01  0.0                     PRAD\n",
       "\n",
       "[3662 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "survival_filtered_feature_df = survival_filtered_feature_df[['sample', 'OS', 'cancer type abbreviation']]\n",
    "display(survival_filtered_feature_df)\n",
    "survival_filtered_feature_df.to_csv(os.path.join(graph_output_folder, 'survival-label.csv'), index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 8.2 Randomize the input label"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Randomize the survival label\n",
    "def input_random(randomized, graph_output_folder):\n",
    "    randomized = False\n",
    "    if randomized == True:\n",
    "        random_survival_filtered_feature_df = survival_filtered_feature_df.sample(frac = 1).reset_index(drop=True)\n",
    "        random_survival_filtered_feature_df.to_csv(os.path.join(graph_output_folder, 'random-survival-label.csv'), index=False)\n",
    "    else:\n",
    "        random_survival_filtered_feature_df = pd.read_csv(os.path.join(graph_output_folder, 'random-survival-label.csv'))\n",
    "    display(random_survival_filtered_feature_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 8.3 Split the randomized input into 5-fold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Split deep learning input into training and test\n",
    "def split_k_fold(k, graph_output_folder):\n",
    "    random_survival_filtered_feature_df = pd.read_csv(os.path.join(graph_output_folder, 'random-survival-label.csv'))\n",
    "    num_points = random_survival_filtered_feature_df.shape[0]\n",
    "    num_div = int(num_points / k)\n",
    "    num_div_list = [i * num_div for i in range(0, k)]\n",
    "    num_div_list.append(num_points)\n",
    "    # Split [random_survival_filtered_feature_df] into [k] folds\n",
    "    for place_num in range(k):\n",
    "        low_idx = num_div_list[place_num]\n",
    "        high_idx = num_div_list[place_num + 1]\n",
    "        print('\\n--------TRAIN-TEST SPLIT WITH TEST FROM ' + str(low_idx) + ' TO ' + str(high_idx) + '--------')\n",
    "        split_input_df = random_survival_filtered_feature_df[low_idx : high_idx]\n",
    "        split_input_df.to_csv(os.path.join(graph_output_folder, 'split-random-survival-label-' + str(place_num + 1) + '.csv'), index=False)\n",
    "        print(split_input_df.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 8.4 Reprocess the edge_index file after loading"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>src</th>\n",
       "      <th>dest</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>405</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>406</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>1331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>1332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>1333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39975</th>\n",
       "      <td>4154</td>\n",
       "      <td>1802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39976</th>\n",
       "      <td>4155</td>\n",
       "      <td>1836</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39977</th>\n",
       "      <td>4156</td>\n",
       "      <td>1864</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39978</th>\n",
       "      <td>4157</td>\n",
       "      <td>1902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39979</th>\n",
       "      <td>4158</td>\n",
       "      <td>1984</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>39980 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        src  dest\n",
       "0         0   405\n",
       "1         0   406\n",
       "2         0  1331\n",
       "3         0  1332\n",
       "4         0  1333\n",
       "...     ...   ...\n",
       "39975  4154  1802\n",
       "39976  4155  1836\n",
       "39977  4156  1864\n",
       "39978  4157  1902\n",
       "39979  4158  1984\n",
       "\n",
       "[39980 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>src</th>\n",
       "      <th>dest</th>\n",
       "      <th>EdgeType</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>405</td>\n",
       "      <td>Gene-Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>406</td>\n",
       "      <td>Gene-Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>1331</td>\n",
       "      <td>Gene-Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>1332</td>\n",
       "      <td>Gene-Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>1333</td>\n",
       "      <td>Gene-Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39975</th>\n",
       "      <td>4154</td>\n",
       "      <td>1802</td>\n",
       "      <td>Gene-Gene-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39976</th>\n",
       "      <td>4155</td>\n",
       "      <td>1836</td>\n",
       "      <td>Gene-Gene-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39977</th>\n",
       "      <td>4156</td>\n",
       "      <td>1864</td>\n",
       "      <td>Gene-Gene-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39978</th>\n",
       "      <td>4157</td>\n",
       "      <td>1902</td>\n",
       "      <td>Gene-Gene-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39979</th>\n",
       "      <td>4158</td>\n",
       "      <td>1984</td>\n",
       "      <td>Gene-Gene-PROT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>39980 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        src  dest        EdgeType\n",
       "0         0   405       Gene-Gene\n",
       "1         0   406       Gene-Gene\n",
       "2         0  1331       Gene-Gene\n",
       "3         0  1332       Gene-Gene\n",
       "4         0  1333       Gene-Gene\n",
       "...     ...   ...             ...\n",
       "39975  4154  1802  Gene-Gene-PROT\n",
       "39976  4155  1836  Gene-Gene-PROT\n",
       "39977  4156  1864  Gene-Gene-PROT\n",
       "39978  4157  1902  Gene-Gene-PROT\n",
       "39979  4158  1984  Gene-Gene-PROT\n",
       "\n",
       "[39980 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>src</th>\n",
       "      <th>dest</th>\n",
       "      <th>EdgeType</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>405</td>\n",
       "      <td>Gene-Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>406</td>\n",
       "      <td>Gene-Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>1331</td>\n",
       "      <td>Gene-Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>1332</td>\n",
       "      <td>Gene-Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>1333</td>\n",
       "      <td>Gene-Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39975</th>\n",
       "      <td>4154</td>\n",
       "      <td>1802</td>\n",
       "      <td>Gene-Gene-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39976</th>\n",
       "      <td>4155</td>\n",
       "      <td>1836</td>\n",
       "      <td>Gene-Gene-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39977</th>\n",
       "      <td>4156</td>\n",
       "      <td>1864</td>\n",
       "      <td>Gene-Gene-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39978</th>\n",
       "      <td>4157</td>\n",
       "      <td>1902</td>\n",
       "      <td>Gene-Gene-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39979</th>\n",
       "      <td>4158</td>\n",
       "      <td>1984</td>\n",
       "      <td>Gene-Gene-PROT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>39980 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        src  dest        EdgeType\n",
       "0         0   405       Gene-Gene\n",
       "1         0   406       Gene-Gene\n",
       "2         0  1331       Gene-Gene\n",
       "3         0  1332       Gene-Gene\n",
       "4         0  1333       Gene-Gene\n",
       "...     ...   ...             ...\n",
       "39975  4154  1802  Gene-Gene-PROT\n",
       "39976  4155  1836  Gene-Gene-PROT\n",
       "39977  4156  1864  Gene-Gene-PROT\n",
       "39978  4157  1902  Gene-Gene-PROT\n",
       "39979  4158  1984  Gene-Gene-PROT\n",
       "\n",
       "[39980 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>src</th>\n",
       "      <th>dest</th>\n",
       "      <th>EdgeType</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>CRK</td>\n",
       "      <td>Gene-Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>CRKL</td>\n",
       "      <td>Gene-Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>NTRK1</td>\n",
       "      <td>Gene-Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>NTRK2</td>\n",
       "      <td>Gene-Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ABL1</td>\n",
       "      <td>NTRK3</td>\n",
       "      <td>Gene-Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39975</th>\n",
       "      <td>SMAD4-PROT</td>\n",
       "      <td>SMAD4</td>\n",
       "      <td>Gene-Gene-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39976</th>\n",
       "      <td>SRC-PROT</td>\n",
       "      <td>SRC</td>\n",
       "      <td>Gene-Gene-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39977</th>\n",
       "      <td>SYK-PROT</td>\n",
       "      <td>SYK</td>\n",
       "      <td>Gene-Gene-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39978</th>\n",
       "      <td>TFRC-PROT</td>\n",
       "      <td>TFRC</td>\n",
       "      <td>Gene-Gene-PROT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39979</th>\n",
       "      <td>TSC1-PROT</td>\n",
       "      <td>TSC1</td>\n",
       "      <td>Gene-Gene-PROT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>39980 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              src   dest        EdgeType\n",
       "0            ABL1    CRK       Gene-Gene\n",
       "1            ABL1   CRKL       Gene-Gene\n",
       "2            ABL1  NTRK1       Gene-Gene\n",
       "3            ABL1  NTRK2       Gene-Gene\n",
       "4            ABL1  NTRK3       Gene-Gene\n",
       "...           ...    ...             ...\n",
       "39975  SMAD4-PROT  SMAD4  Gene-Gene-PROT\n",
       "39976    SRC-PROT    SRC  Gene-Gene-PROT\n",
       "39977    SYK-PROT    SYK  Gene-Gene-PROT\n",
       "39978   TFRC-PROT   TFRC  Gene-Gene-PROT\n",
       "39979   TSC1-PROT   TSC1  Gene-Gene-PROT\n",
       "\n",
       "[39980 rows x 3 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "graph_output_folder = 'graph-data'\n",
    "form_data_path = './' + graph_output_folder + '/form_data'\n",
    "edge_index = np.load(form_data_path + '/edge_index.npy')\n",
    "# Convert the 2D array into a DataFrame\n",
    "edge_index_df = pd.DataFrame(edge_index.T, columns=['src', 'dest'])\n",
    "\n",
    "gene_edge_num_df = pd.read_csv(os.path.join(graph_output_folder, 'all-gene-edge-num.csv'))\n",
    "src_gene_list = list(gene_edge_num_df['src'])\n",
    "dest_gene_list = list(gene_edge_num_df['dest'])\n",
    "edgetype_list = list(gene_edge_num_df['EdgeType'])\n",
    "gene_edge_num_reverse_df = pd.DataFrame({'src': dest_gene_list, 'dest': src_gene_list, 'EdgeType': edgetype_list})\n",
    "gene_edge_num_all_df = pd.concat([gene_edge_num_df, gene_edge_num_reverse_df]).drop_duplicates().sort_values(by=['src', 'dest']).reset_index(drop=True)\n",
    "\n",
    "display(edge_index_df)\n",
    "display(gene_edge_num_all_df)\n",
    "merged_gene_edge_num_all_df = pd.merge(gene_edge_num_all_df, edge_index_df, on=['src', 'dest'], how='inner')\n",
    "display(merged_gene_edge_num_all_df)\n",
    "merged_gene_edge_num_all_df.to_csv(os.path.join(graph_output_folder, 'merged-gene-edge-num-all.csv'), index=False)\n",
    "\n",
    "merged_gene_edge_name_all_df = merged_gene_edge_num_all_df.replace(sorted_all_gene_dict)\n",
    "display(merged_gene_edge_name_all_df)\n",
    "merged_gene_edge_name_all_df.to_csv(os.path.join(graph_output_folder, 'merged-gene-edge-name-all.csv'), index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "win_env",
   "language": "python",
   "name": "win_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
